### PR TITLE
Update Rustc

### PR DESCRIPTION
**Pull request by xldenis**
This time against the right base! Theres a few final points to deal with in the metadata. I've based myself of how I did these things in Creusot when I did a similar update.

### CODE REVIEW

File: prusti-interface/src/specs/encoder.rs [hunk 1]
```diff
@@ -59,47 +79,76 @@ impl<'tcx> Encoder for DefSpecsEncoder<'tcx> {
...
+    }
+    fn encode_crate_num(&mut self, cnum: CrateNum) {
+        self.tcx.stable_crate_id(cnum).encode(self)
+    }
+    fn encode_def_index(&mut self, _: DefIndex) {
+        panic!("encoding `DefIndex` without context");
+    }
+    fn encode_def_id(&mut self, id: DefId) {
+        self.tcx.def_path_hash(id).encode(self)
+    }
```
Comment (xldenis) [inline prusti-interface/src/specs/encoder.rs]: @Aurel300 If you're interested, in Creusot we have all this logic extracted in a generic, creusot independent manner in a standalone crate. we could move that to a separate repo and reuse it in Prusti directly. This stuff is annoying and actually rather tricky to get 100% correct.

**Comment by xldenis** [comment]
I think there are still a few minor touch-ups and fixups, but the code works and compiles as far as I can tell.

File: prusti-interface/src/specs/encoder.rs (hunk 1 above)
Comment (Aurel300) [reply prusti-interface/src/specs/encoder.rs]: Sounds good, which crate is it?

File: mir-state-analysis/src/free_pcs/impl/triple.rs [hunk 2]
```diff
@@ -98,6 +98,9 @@ impl<'tcx> Visitor<'tcx> for Fpcs<'_, 'tcx> {
                 self.requires_write(place);
                 self.ensures_write(place);
             }
+            // Based off def of `Call`
+            &TailCall { .. } => {}
```
Comment (Aurel300) [inline mir-state-analysis/src/free_pcs/impl/triple.rs]: I'm guessing this should also do whatever `Return` does?

File: prusti-encoder/src/encoders/type/predicate.rs [hunk 3]
```diff
@@ -1,5 +1,6 @@
 use prusti_rustc_interface::{
-    abi,
+    // abi,
+    target::abi,
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/predicate.rs:2]: Why the `// abi` comment? Also in `type/domain.rs`.

**Comment by Aurel300** [comment]
Otherwise looks good. Re: metadata, we should make sure the changes work for (cross-crate) extern specs (unsure if we have tests for that in v2).

File: prusti-interface/src/specs/encoder.rs (hunk 1 above)
Comment (xldenis) [reply prusti-interface/src/specs/encoder.rs]: It's called `creusot-metadata` in our repo currently

File: prusti-encoder/src/encoders/type/predicate.rs (hunk 3 above)
Comment (xldenis) [reply prusti-encoder/src/encoders/type/predicate.rs:2]: They moved stuff around, the relevant types are no longer exported as part of `abi` but `target::abi`.

**Comment by xldenis** [comment]
> Re: metadata, we should make sure the changes work for (cross-crate) extern specs (unsure if we have tests for that in v2).

Yea, I'll need to test that today before merge. However, if/when we switch to `creusot-metadata` (which I can extract into a `custom-rustc-metadata` crate in a standalone repo, I know the results are correct there (by virtue of testing in creusot).

**Comment by xldenis** [comment]
Ok, its rather hard to be definitive until more of the prusti translation is implemented, but running:

```
 ./x.py build --all && \
RUST_BACKTRACE=1 PRUSTI_DUMP_VIPER_PROGRAM=true PRUSTI_DUMP_DEBUG_INFO=true \
 ./x.py ++verbose verify-test prusti-tests/tests/verify/pass/foo.rs
```

seems to work for randomly chosen files, where by "work" i mean "same error as `rewrite-2023`.

**Pull request merged**